File: 035_search_insert_position.java
iclass Solution {
    public int searchInsert(int[] nums, int target) {
        int low = 0, high = nums.length - 1;
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (nums[mid] > target) {
                high = mid;
            } else if (nums[mid] < target){
                low = mid + 1;
            } else {
                low = mid;
                break;
            }
        }
        while (low > 1 && nums[low] == nums[low - 1]) {
            low--;
        }
        return (nums[low] < target) ? nums.length: low;
        // int result = -1;
        // for (int i = 0; i < nums.length; i++){
        //     if (target <= nums[i]){
        //         result = i;
        //         break;
        //     }
        // }
        // if (result == -1) result = nums.length;
        // return result;
    }
}

File: python/035_search_insert_position.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        # use binary search to find target
        lo, hi = 0, len(nums) - 1
        while (lo < hi):
            mid = int((lo + hi) / 2)
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                lo = mid + 1
            else:
                hi = mid;
        if nums[lo] < target:
            return lo + 1
        return lo
